subgraphs/compound-forks/protocols/dforce/src/mapping.ts

```diff
@@ -704,3 +551,166 @@ export function getOrCreateMarketStatus(marketId: string): _DforceMarketStatus {
+  event: DistributionSupplySpeedUpdated
+): void {
+  const market = Market.load(event.params.iToken.toHexString());
+  if (!market) {
+    log.error("Market not found for address {}", [
+      event.params.iToken.toHexString(),
+    ]);
+    return;
+  }
+  const emissionsAmount = market.rewardTokenEmissionsAmount!;
+  emissionsAmount[1] = BigDecimalTruncateToBigInt(
+    event.params.supplySpeed.toBigDecimal().times(blocksPerDay)
+  );
+  market.rewardTokenEmissionsAmount = emissionsAmount;
+  market.save();
+
+  updateRewardPrices(event.params.iToken);
+}
+
+function initRewards(marketAddr: Address): void {
+  const market = Market.load(marketAddr.toHexString());
+  if (!market) {
+    log.error("Market not found for address {}", [marketAddr.toHexString()]);
+    return;
+  }
+
+  market.rewardTokens = getOrCreateRewardTokens();
+  market.rewardTokenEmissionsAmount = [BIGINT_ZERO, BIGINT_ZERO];
+  market.rewardTokenEmissionsUSD = [BIGDECIMAL_ZERO, BIGDECIMAL_ZERO];
+  market.save();
+}
+
+// initially create reward tokens
+// reward is always DForce token
+function getOrCreateRewardTokens(): string[] {
+  let token = Token.load(rewardTokenAddr.toHexString());
+  if (token == null) {
+    token = new Token(rewardTokenAddr.toHexString());
+    // special handle for DF token since its name and symbol is byte32
+    if (rewardTokenAddr == DF_ADDRESS) {
+      token.name = "dForce";
+      token.symbol = "DF";
+      token.decimals = 18;
+    } else {
+      const tokenContract = ERC20.bind(rewardTokenAddr);
+      token.name = getOrElse<string>(tokenContract.try_name(), "unknown");
+      token.symbol = getOrElse<string>(tokenContract.try_symbol(), "unknown");
+      token.decimals = getOrElse<i32>(tokenContract.try_decimals(), 18);
+    }
+    token.save();
+  }
+
+  const borrowRewardTokenId = RewardTokenType.BORROW.concat("-").concat(
+    token.id
+  );
+  let borrowRewardToken = RewardToken.load(borrowRewardTokenId);
+  if (borrowRewardToken == null) {
+    borrowRewardToken = new RewardToken(borrowRewardTokenId);
+    borrowRewardToken.token = token.id;
+    borrowRewardToken.type = RewardTokenType.BORROW;
+    borrowRewardToken.save();
+  }
+
+  const depositRewardTokenId = RewardTokenType.DEPOSIT.concat("-").concat(
+    token.id
+  );
+  let depositRewardToken = RewardToken.load(depositRewardTokenId);
+  if (depositRewardToken == null) {
+    depositRewardToken = new RewardToken(depositRewardTokenId);
+    depositRewardToken.token = token.id;
+    depositRewardToken.type = RewardTokenType.DEPOSIT;
+    depositRewardToken.save();
+  }
+
+  return [borrowRewardTokenId, depositRewardTokenId];
+}
+
+function updateRewardPrices(marketAddress: Address): void {
+  const market = Market.load(marketAddress.toHexString());
+  if (!market) {
+    log.error("Market not found for address {}", [marketAddress.toHexString()]);
+    return;
+  }
+  const rewardToken = RewardToken.load(market.rewardTokens![0]);
+  if (!rewardToken) {
```

Comment:
This is unlikely to happend, but if it does, should we call `getOrCreateRewardTokens` fix the error?

Reply:
I guess if this isn't there then there are likely more issues. So I guess I would want to throw an error and quit.. What do you think?

Reply:
Since this is a very common occurence when we build subgraphs we can try to come up with some standard decision. I think the options are:
- Log an error and skip the calculation that follows
  - good: the subgraph doesn’t break, keeps indexing
  - bad: no one looks at errors, and if we do, they get buried in other logs. Might never get fixed unless it affects data quality noticeably 
- Have some fallback or use `getOrCreate` to ensure the entity is always there
  - good: the subgraph doesn’t break and keeps indexing
  - bad: if the entity was supposed to be there and wasn’t, we won’t even log an error. Might be affecting something else somewhere and make it harder to debug when we create the entity “again”.
- Fail with a critical error, to make sure someone comes and fix the underlying issue
  - good: we ensure someone will take a look at it
  - bad: subgraph stops syncing until fixed. Whomever comes to fix it might just default to any of the above options, adding one step.

Most of this I think will be fixed with the libraries we build for each schema though. And maybe, instead of this, we can think on what each log level means. If we purely keep errors logs at a minimum, only logging what really signals that something needs to get fixed, together with the log persistence tool we could both know that an error happened and keep the subgraph going while we fix it. Will bring up this discussion on an issue or discord to not block this PR.

Getting back to the topic at hand 😅 I don’t like any of the options at hand, but I prefer the `log an error and return` option too.

Reply:
makes sense.